Support mounting host volumes in build container

actgardner opened the pull request:
Add support for arbitrary volume mounts in the build container. This is necessary to support cases like mounting an ssh-agent socket so a Golang build can pull private images, for instance. In general this could be useful to expose build-time secrets.

simonjjones commented:
Hi @actgardner, thanks again for your PR.

I have some concerns around how allowing the mounting of arbitrary host volumes would allow pack consumers to contravene our platform specification, configuring mounts that could break [compatibility guarantees](https://github.com/buildpack/spec/blob/master/platform.md#compatibility-guarantees) by overwriting components in the build image with non-ABI-compatible binaries. If a pack consumer were to mount volumes in such a way, this could lead to ABI breakages between build-time and runtime. It could also lead to built images varying when built in different environments, despite configuring the same inputs, challenging expectations that they would be consistent. I would also note that while [docker’s comparable `--secret` option](https://docs.docker.com/develop/develop-images/build_enhancements/) does allow arbitrary mounting, it only allows a single file, making it less likely that consumers will replace a substantial part of the image.

I am, however, very interested in helping you solve your specific problem.

The [buildpack specification](https://github.com/buildpack/spec/blob/master/buildpack.md#detection) has a provision for platform specific extensions to exist in the /platform/ folder. If mounting the SSH agent socket within `/platform/` would solve your particular problem then you could update your PR in either of the following ways:

- Document the requirement for mounted volumes to be prefixed with `/platform/` AND reject any volumes configured without that prefix
- Prefix all configured volumes with `/platform/` AND ensure that this alteration is suitably discoverable and documented to avoid unexpected behaviour

I would also consider a more specific change, in which we didn’t try to make this generic at all, but instead focused on the best way to share SSH authorization with the build container. Docker’s approach of [forwarding your ssh-agent with just the `-ssh` flag](https://medium.com/@tonistiigi/build-secrets-and-ssh-forwarding-in-docker-18-09-ae8161d066) could be more intuitive, in this possibly common use case.

I’d be happy to follow up further over a video call if that would help to expand on my concerns, or if you’d like to discuss the alternatives in more detail. You’re also welcome to reach out to the broader team in our [slack](https://slack.buildpacks.io/). We also have two weekly working group meetings on Wednesday at 3pm ET and Thursday at 2pm ET where contributors and maintainers discuss PRs under review amongst other things. Let me know if you’d like an invite.

ameyer-pivotal commented:
Hi @actgardner,

After some discussion, the team is in favor of either:

1. Prefix all configured volumes with `/platform/`. For example, something like `--mount=a:b` => `/platform/b`. We could take care of the documentation of such feature.

2. The more specific approach of something like an `-ssh` flag. This might require additional discussion.

If you'd like to go with option 1, would you like to adjust this PR or may we close it and expect another? As always, thanks for the contribution! :)

actgardner commented:
Hey @ameyer-pivotal, sorry for the slow response. Option 1 sounds good to me - we need SSH sockets but also build-time secrets so this kills two birds with one stone. I should have some bandwidth to freshen this PR up for next week.  Thanks for your feedback!

actgardner commented:
What's this project's attitude towards bringing in code from Moby? This currently works on Linux but parsing the volume mount syntax for Windows is extremely involved: https://github.com/moby/moby/blob/master/volume/mounts/windows_parser.go

ameyer-pivotal commented on internal/commands/build.go:
```diff
@@ -117,6 +119,7 @@ func buildCommandFlags(cmd *cobra.Command, buildFlags *BuildFlags, cfg config.Co
 	cmd.Flags().StringSliceVarP(&buildFlags.Buildpacks, "buildpack", "b", nil, "Buildpack reference in the form of '<buildpack>@<version>',\n  path to a buildpack directory (not supported on Windows),\n  path/URL to a buildpack .tar or .tgz file, or\n  the name of a packaged buildpack image"+multiValueHelp("buildpack"))
 	cmd.Flags().StringVar(&buildFlags.Network, "network", "", "Connect detect and build containers to network")
 	cmd.Flags().StringVarP(&buildFlags.DescriptorPath, "descriptor", "d", "", "Path to the project descriptor file")
+	cmd.Flags().StringArrayVar(&buildFlags.Volumes, "volume", nil, "Mount host volume into the build container")
```
Should append `+multiValueHelp("volume")` here

See: https://github.com/buildpacks/pack/pull/400/files#diff-c8fbe25a03e6f0fd74a18df1bc72bb68R119 above

ameyer-pivotal commented on build.go:
```diff
@@ -522,3 +529,16 @@ func randString(n int) string {
 	return string(b)
 }
+
+func buildPlatformVolumes(volumes []string) ([]string, error) {
```
I'm wondering if we could skip any translation logic and just pass the strings as-is from the CLI flag to the underlying docker client? Wouldn't any formatting errors be surfaced at that time? If that works, it could simplify the implementation of this feature quite a bit.

ameyer-pivotal replied:
Oh, I forgot that we need to prepend `platform` to the in-container mount location...

ameyer-pivotal commented on internal/build/phase_test.go:
```diff
@@ -313,6 +313,20 @@ func testPhase(t *testing.T, when spec.G, it spec.S) {
 				})
 			})
+
+			when("#WithPlatformVolumes", func() {
```
Is this test incomplete, or am I missing something? :)

actgardner replied:
Whoops! This was a mistake, the tests for the platform prefix should all live in `build_test.go`

actgardner commented:
This currently works for Linux and Windows host machines building Linux containers. If we want to add support for Windows containers we'll need to add an extra flag specifying that the container is Windows so we know what parsing logic to use (unless there's a trick I'm missing where we can detect the container OS automatically).

ameyer-pivotal commented:
> This currently works for Linux and Windows host machines building Linux containers. If we want to add support for Windows containers we'll need to add an extra flag specifying that the container is Windows so we know what parsing logic to use (unless there's a trick I'm missing where we can detect the container OS automatically).

Thanks @actgardner. I think we can punt on Windows containers for now as long as it works for Windows host machines. We have other work to do to enable Windows containers with CNBs.

cc: @micahyoung, @jromero

jromero commented:
Hi @actgardner,

Overall great work! Here are a few comments from running manual acceptance on this PR.

### ❗️Platform Read-only

It dawned on us that the default for a volume mount would be `rw`.  After looking into the [spec](https://github.com/buildpacks/spec/blob/master/buildpack.md#build) and [consulting with others](https://buildpacks.slack.com/archives/CCP60GJLS/p1582572398004400), we came to the conclusion that this feature should only allow for a read-only mount. If you have any concerns or feedback with that decision please let us know.

##### Expected

`pack` users should not be able to provide a file mode to the `--volume` parameter. We would then expect that the filemode for container creation is hardcoded to `ro`.

-----

### ❗️ Unclear error when source dir is relative

```shell
$ ./out/pack build some-app -B cnbs/sample-builder:alpine -b ../samples/buildpacks/hello-world/ --volume non-existant:/ptest2
alpine: Pulling from cnbs/sample-builder
Digest: sha256:cc0fe2eb24295b7f91780c96677eced421b307693686a825a1d33b9bb10107e8
Status: Image is up to date for cnbs/sample-builder:alpine
alpine: Pulling from cnbs/sample-stack-run
Digest: sha256:ec8f7e2edbcf65185a36b51a5c88975964a78a1e3d5fb84d618e2720c25f110d
Status: Image is up to date for cnbs/sample-stack-run:alpine
===> DETECTING
[detector] io.buildpacks.samples.hello-world 0.0.1
===> RESTORING
===> ANALYZING
===> BUILDING
ERROR: failed to create 'builder' container: Error response from daemon: invalid volume specification: ':/platform/ptest2:'
```

-----

### ❓ Help text should show format

Without knowing the details of the implementation it's unknown what format is expected. Specially, given the fact that it won't be able to follow exactly what `docker -v` accepts due to filemode not being an option. See `--env` for a sample of a flag that provides format information.

##### Expected

The help text should indicate the format of `source:target`.

-----

### ❓ Confusing absolute path request

Given that the target path is actually nested under `/platform` it seems confusing to request that the path be absolute which isn't actually _the_ final absolute path.

```
$ ./out/pack build some-app -B cnbs/sample-builder:alpine -b ../samples/buildpacks/hello-world/ --volume /tmp/ptest2:ptest2
....
ERROR: Platform volume "/tmp/ptest2:ptest2" has invalid format: invalid volume specification: '/tmp/ptest2:ptest2': invalid mount config for type "bind": invalid mount path: 'ptest2' mount path must be absolute
```

##### Expected

This error should not show up. The expectation is that the following syntax works with or without a leading file path separator. For example:

- `/tmp/ptest2:ptest2` => `/platform/ptest2`
- `/tmp/ptest2:/ptest2` => `/platform/ptest2`

-----

Other things we looked at...

### ✅ Happy path

```shell
./out/pack build some-app -B cnbs/sample-builder:alpine -b ../samples/buildpacks/hello-world/ --volume /tmp/ptest:/ptest
alpine: Pulling from cnbs/sample-builder
Digest: sha256:cc0fe2eb24295b7f91780c96677eced421b307693686a825a1d33b9bb10107e8
Status: Image is up to date for cnbs/sample-builder:alpine
alpine: Pulling from cnbs/sample-stack-run
Digest: sha256:ec8f7e2edbcf65185a36b51a5c88975964a78a1e3d5fb84d618e2720c25f110d
Status: Image is up to date for cnbs/sample-stack-run:alpine
===> DETECTING
[detector] io.buildpacks.samples.hello-world 0.0.1
===> RESTORING
===> ANALYZING
===> BUILDING
[builder] ---> Hello World buildpack
[builder] Platform dir contents: /platform
[builder] /platform:
[builder] total 12
[builder] drwxr-xr-x    1 root     root          4096 Feb 24 20:22 .
[builder] drwxr-xr-x    1 root     root          4096 Feb 24 20:22 ..
[builder] drwxr-xr-x    1 root     root          4096 Jan  1  1980 env
[builder] drwxr-xr-x    3 cnb      cnb             96 Feb 24 18:49 ptest
[builder]
[builder] /platform/env:
[builder] total 8
[builder] drwxr-xr-x    1 root     root          4096 Jan  1  1980 .
[builder] drwxr-xr-x    1 root     root          4096 Feb 24 20:22 ..
[builder]
[builder] /platform/ptest:
[builder] total 4
[builder] drwxr-xr-x    3 cnb      cnb             96 Feb 24 18:49 .
[builder] drwxr-xr-x    1 root     root          4096 Feb 24 20:22 ..
[builder] -rw-r--r--    1 cnb      cnb              0 Feb 24 18:49 file1
[builder]
[builder] ---> Done
===> EXPORTING
[exporter] Reusing layer 'app'
[exporter] Reusing layer 'config'
[exporter] Reusing layer 'launcher'
[exporter] *** Images (24344139c149):
[exporter]       index.docker.io/library/some-app:latest
===> CACHING
Successfully built image some-app
```

### ✅ Prevents path traversal

```shell
./out/pack build some-app -B cnbs/sample-builder:alpine -b ../samples/buildpacks/hello-world/ --volume /tmp/ptest2:/../layers/ptest2
alpine: Pulling from cnbs/sample-builder
Digest: sha256:cc0fe2eb24295b7f91780c96677eced421b307693686a825a1d33b9bb10107e8
Status: Image is up to date for cnbs/sample-builder:alpine
alpine: Pulling from cnbs/sample-stack-run
Digest: sha256:ec8f7e2edbcf65185a36b51a5c88975964a78a1e3d5fb84d618e2720c25f110d
Status: Image is up to date for cnbs/sample-stack-run:alpine
===> DETECTING
[detector] io.buildpacks.samples.hello-world 0.0.1
===> RESTORING
===> ANALYZING
===> BUILDING
[builder] ---> Hello World buildpack
[builder] Platform dir contents: /platform
[builder] /platform:
[builder] total 16
[builder] drwxr-xr-x    1 root     root          4096 Feb 24 20:05 .
[builder] drwxr-xr-x    1 root     root          4096 Feb 24 20:05 ..
[builder] drwxr-xr-x    1 root     root          4096 Jan  1  1980 env
[builder] drwxr-xr-x    3 root     root          4096 Feb 24 20:05 layers
[builder]
[builder] /platform/env:
[builder] total 8
[builder] drwxr-xr-x    1 root     root          4096 Jan  1  1980 .
[builder] drwxr-xr-x    1 root     root          4096 Feb 24 20:05 ..
[builder]
[builder] /platform/layers:
[builder] total 8
[builder] drwxr-xr-x    3 root     root          4096 Feb 24 20:05 .
[builder] drwxr-xr-x    1 root     root          4096 Feb 24 20:05 ..
[builder] drwxr-xr-x    3 cnb      cnb             96 Feb 24 19:13 ptest2
[builder]
[builder] /platform/layers/ptest2:
[builder] total 8
[builder] drwxr-xr-x    3 cnb      cnb             96 Feb 24 19:13 .
[builder] drwxr-xr-x    3 root     root          4096 Feb 24 20:05 ..
[builder] -rw-r--r--    1 cnb      cnb              6 Feb 24 19:13 new-file
[builder]
[builder] ---> Done
```

```shell
./out/pack build some-app -B cnbs/sample-builder:alpine -b ../samples/buildpacks/hello-world/ --volume /tmp/ptest:/ptest/../../layers/ptest
alpine: Pulling from cnbs/sample-builder
Digest: sha256:cc0fe2eb24295b7f91780c96677eced421b307693686a825a1d33b9bb10107e8
Status: Image is up to date for cnbs/sample-builder:alpine
alpine: Pulling from cnbs/sample-stack-run
Digest: sha256:ec8f7e2edbcf65185a36b51a5c88975964a78a1e3d5fb84d618e2720c25f110d
Status: Image is up to date for cnbs/sample-stack-run:alpine
===> DETECTING
[detector] io.buildpacks.samples.hello-world 0.0.1
===> RESTORING
===> ANALYZING
===> BUILDING
[builder] ---> Hello World buildpack
[builder] Platform dir contents: /platform
[builder] /platform:
[builder] total 16
[builder] drwxr-xr-x    1 root     root          4096 Feb 24 20:24 .
[builder] drwxr-xr-x    1 root     root          4096 Feb 24 20:24 ..
[builder] drwxr-xr-x    1 root     root          4096 Jan  1  1980 env
[builder] drwxr-xr-x    3 root     root          4096 Feb 24 20:24 layers
[builder]
[builder] /platform/env:
[builder] total 8
[builder] drwxr-xr-x    1 root     root          4096 Jan  1  1980 .
[builder] drwxr-xr-x    1 root     root          4096 Feb 24 20:24 ..
[builder]
[builder] /platform/layers:
[builder] total 8
[builder] drwxr-xr-x    3 root     root          4096 Feb 24 20:24 .
[builder] drwxr-xr-x    1 root     root          4096 Feb 24 20:24 ..
[builder] drwxr-xr-x    3 cnb      cnb             96 Feb 24 18:49 ptest
[builder]
[builder] /platform/layers/ptest:
[builder] total 4
[builder] drwxr-xr-x    3 cnb      cnb             96 Feb 24 18:49 .
[builder] drwxr-xr-x    3 root     root          4096 Feb 24 20:24 ..
[builder] -rw-r--r--    1 cnb      cnb              0 Feb 24 18:49 file1
[builder]
[builder] ---> Done
```

actgardner commented:
@jromero Thanks for the feedback :) I believe I've addressed most of your feedback:

- **Platform Read-only** - we now strip off any options and append `:ro`
- **Unclear error when source dir is relative** - this is actually the docker syntax to mount a named volume, not a relative path. The volume name is now correctly passed through to the builder.
- **Help text should show format** - the help text now shows example syntax

**Confusing absolute path request** - this is difficult to handle in the general case. We rely on the Docker codebase to parse the mount syntax, and the parser will error out if the target path isn't absolute. If this is too much of a limitation, we could conceivably require users to provide the `/platform` absolute path themselves and make it an error if they try to mount a volume elsewhere? It's not clear to me if that provides a better experience.

natalieparellano commented:
Hey @actgardner this looks great! I'm attaching for posterity the results of my validation. cc @jromero 

Happy path and “prevents path traversal” test cases are still passing.

Modified the hello-world buildpack to attempt to write a file in the mounted volume, tried to mount the volume as read-write, and verified that the build still failed:

```
$  ./out/pack build some-app -B cnbs/sample-builder:alpine -b ../samples/buildpacks/hello-world/ --volume/tmp/ptest:/ptest:rw
alpine: Pulling from cnbs/sample-builder
Digest: sha256:b0e33eee55ee7544812e217ea8287b945e721a5bce66505cca83a189ff972f57
Status: Image is up to date for cnbs/sample-builder:alpine
alpine: Pulling from cnbs/sample-stack-run
Digest: sha256:f1bef0812ec1763345774ed6566c6db97893e24aa88e8fd00576c745f17f30cb
Status: Image is up to date for cnbs/sample-stack-run:alpine
Warning: buildpack io.buildpacks.samples.hello-world@0.0.1 already exists on builder and will be overwritten
===> DETECTING
[detector] io.buildpacks.samples.hello-world 0.0.1
===> ANALYZING
===> RESTORING
===> BUILDING
[builder] ---> Hello World buildpack
[builder] Platform dir contents: /platform
[builder] /platform:
[builder] total 12
[builder] drwxr-xr-x    1 root     root          4096 Feb 25 16:13 .
[builder] drwxr-xr-x    1 root     root          4096 Feb 25 16:13 ..
[builder] drwxr-xr-x    1 root     root          4096 Jan  1  1980 env
[builder] drwxr-xr-x    3 cnb      cnb             96 Feb 25 16:07 ptest
[builder] 
[builder] /platform/env:
[builder] total 8
[builder] drwxr-xr-x    1 root     root          4096 Jan  1  1980 .
[builder] drwxr-xr-x    1 root     root          4096 Feb 25 16:13 ..
[builder] 
[builder] /platform/ptest:
[builder] total 4
[builder] drwxr-xr-x    3 cnb      cnb             96 Feb 25 16:07 .
[builder] drwxr-xr-x    1 root     root          4096 Feb 25 16:13 ..
[builder] -rw-r--r--    1 cnb      cnb              0 Feb 25 16:07 file1
[builder] 
[builder] touch: /platform/ptest/file2: Read-only file system
[builder] ERROR: failed to build: exit status 1
ERROR: failed with status code: 7
```

Tried to mount a named volume, and verified that it works.

Tried to mount a non-existent named volume; it appears that current implementation will create a volume and mount it at the target:
```
$  ./out/pack build some-app -B cnbs/sample-builder:alpine -b ../samples/buildpacks/hello-world/ --volume non-existant:/ptest2
alpine: Pulling from cnbs/sample-builder
Digest: sha256:b0e33eee55ee7544812e217ea8287b945e721a5bce66505cca83a189ff972f57
Status: Image is up to date for cnbs/sample-builder:alpine
alpine: Pulling from cnbs/sample-stack-run
Digest: sha256:f1bef0812ec1763345774ed6566c6db97893e24aa88e8fd00576c745f17f30cb
Status: Image is up to date for cnbs/sample-stack-run:alpine
Warning: buildpack io.buildpacks.samples.hello-world@0.0.1 already exists on builder and will be overwritten
===> DETECTING
[detector] io.buildpacks.samples.hello-world 0.0.1
===> ANALYZING
===> RESTORING
===> BUILDING
[builder] ---> Hello World buildpack
[builder] Platform dir contents: /platform
[builder] /platform:
[builder] total 16
[builder] drwxr-xr-x    1 root     root          4096 Feb 25 16:20 .
[builder] drwxr-xr-x    1 root     root          4096 Feb 25 16:20 ..
[builder] drwxr-xr-x    1 root     root          4096 Jan  1  1980 env
[builder] drwxr-xr-x    2 root     root          4096 Feb 25 16:18 ptest2
[builder] 
[builder] /platform/env:
[builder] total 8
[builder] drwxr-xr-x    1 root     root          4096 Jan  1  1980 .
[builder] drwxr-xr-x    1 root     root          4096 Feb 25 16:20 ..
[builder] 
[builder] /platform/ptest2:
[builder] total 8
[builder] drwxr-xr-x    2 root     root          4096 Feb 25 16:18 .
[builder] drwxr-xr-x    1 root     root          4096 Feb 25 16:20 ..
[builder] 
[builder] ---> Done
===> EXPORTING
[exporter] Reusing layer 'launcher'
[exporter] Reusing 1/1 app layer(s)
[exporter] Reusing layer 'config'
[exporter] *** Images (5230e214aa4e):
[exporter]       index.docker.io/library/some-app:latest
Successfully built image some-app
```
```
$  docker volume list
DRIVER              VOLUME NAME
...
local               non-existant
```

Verified that help text shows expected format for specifying mounts:
```
$  ./out/pack build --help
...
      --volume stringArray     Mount host volume into the build container, in the form '<host path>:<target path>'. Target path will be prefixed with '/platform/'
```
